apps/openmw/mwmechanics/spellcasting.cpp

```diff
@@ -1009,18 +1016,72 @@ namespace MWMechanics
 
             MWRender::Animation* animation = MWBase::Environment::get().getWorld()->getAnimation(mCaster);
 
-            if (animation && mCaster.getClass().isActor()) // TODO: Non-actors should also create a spell cast vfx even if they are disabled (animation == NULL)
+            const ESM::Static* castStatic;
+
+            if (!effect->mCasting.empty())
+                castStatic = store.get<ESM::Static>().find (effect->mCasting);
+            else
+                castStatic = store.get<ESM::Static>().find ("VFX_DefaultCast");
+
+            // check if the effect was already added
+            if (std::find(addedEffects.begin(), addedEffects.end(), "meshes\\" + castStatic->mModel) != addedEffects.end())
+                continue;
+
+            std::string texture = effect->mParticle;
+
+            float scale = 1.0f;
+            osg::Vec3f pos (mCaster.getRefData().getPosition().asVec3());
+
+            if (animation && mCaster.getClass().isNpc())
             {
-                const ESM::Static* castStatic;
+                // For NOC we should take race height as scaling factor
+                const ESM::NPC *npc = mCaster.get<ESM::NPC>()->mBase;
+                const MWWorld::ESMStore &esmStore =
+                    MWBase::Environment::get().getWorld()->getStore();
 
-                if (!effect->mCasting.empty())
-                    castStatic = store.get<ESM::Static>().find (effect->mCasting);
-                else
-                    castStatic = store.get<ESM::Static>().find ("VFX_DefaultCast");
+                const ESM::Race *race =
+                    esmStore.get<ESM::Race>().find(npc->mRace);
 
-                std::string texture = effect->mParticle;
+                scale = npc->isMale() ? race->mData.mHeight.mMale : race->mData.mHeight.mFemale;
+            }
+            else
+            {
+                std::string casterModel = mCaster.getClass().getModel(mCaster);
+                osg::ref_ptr<osg::Node> model = MWBase::Environment::get().getWorld()->getInstance(casterModel);
```

Comment:
As I said before we don't need to create an instance just to get the bounds of the model, that's wasted resources. Template is sufficient. Try something like this:
```
osg::ref_ptr<const osg::Node> node = sceneMgr->getTemplate(name);
computeBoundsVisitor.accept(const_cast<osg::Node*>(node.get()));
```
The const_cast is unfortunate, but I'll prefer it over a loss of performance.